### PR TITLE
Allow build DALI for different architectures as well

### PR DESCRIPTION
**Pull request by JanuszL**
- allows specification of DALI architecture as a build parameter,
  currently only x86_64 is supported
- allows passing CUDA_TARGET_ARCHS as a build parameter

Signed-off-by: Janusz Lisiecki <jlisiecki@nvidia.com>

#### Why we need this PR?
- allow build DALI for different architectures as well

#### What happened in this PR?
 - allows specification of DALI architecture as a build parameter,  currently only x86_64 is supported
 - allows passing CUDA_TARGET_ARCHS as a build parameter
 - tested in CI

**JIRA TASK**: [DALI-1116]

### CODE REVIEW

File: tools/test_bundled_libs.py [hunk 1]
```diff
@@ -18,6 +18,14 @@
 from sys import argv
 import subprocess
 
+# Check if any element in the elms list matches the value
+def get_list_elm_match(value, elms):
```
Comment (jantonguirao) [inline tools/test_bundled_libs.py]:
```
def check_list_elm_match(value, elms):
    return any(e in value for e in elms)
```

Comment (JanuszL) [reply tools/test_bundled_libs.py]: Done

File: docs/compilation.rst [hunk 2]
```diff
@@ -30,14 +30,17 @@ Change directory (``cd``) into ``docker`` directory and run ``./build.sh``. If n
 * CUDA_VERSION - CUDA toolkit version (9 for 9.0 or 10 for 10.0). Default is ``10``. If the version is prefixed with `.` then any value ``XX`` can be passed and the user needs to make sure that Dockerfile.cudaXX.deps is present in `docker/` directory.
 * NVIDIA_BUILD_ID - Custom ID of the build. Default is ``1234``.
 * CREATE_WHL - Create a standalone wheel. Default is ``YES``.
-* BUILD_TF_PLUGIN - Create a DALI TensorFlow plugin as well. Default is ``NO``.
+* BUILD_TF_PLUGIN - Create a DALI TensorFlow plugin wheel as well. Default is ``NO``.
+* PREBUILD_TF_PLUGINS - If prebuild DALI TensorFlow plugin. Default is ``YES``.
```
Comment (awolant) [inline docs/compilation.rst]: Maybe a bit more info on how these two options interact would be good here?

Comment (JanuszL) [reply docs/compilation.rst]: Done

File: docs/compilation.rst [hunk 3]
```diff
@@ -26,18 +26,21 @@ Building Python wheel and (optionally) Docker image
...
+* CUDA_VERSION - CUDA toolkit version (9 for 9.0 or 10 for 10.0). The default is ``10``. If the version is prefixed with `.` then any value ``XX`` can be passed and the user needs to make sure that Dockerfile.cudaXX.deps is present in `docker/` directory.
+* NVIDIA_BUILD_ID - Custom ID of the build. The default is ``1234``.
+* CREATE_WHL - Create a standalone wheel. The default is ``YES``.
+* BUILD_TF_PLUGIN - Create a DALI TensorFlow plugin wheel as well. The default is ``NO``.
+* PREBUILD_TF_PLUGINS - If prebuild DALI TensorFlow plugin. It should be used together with BUILD_TF_PLUGIN option. If both options are set to ``YES`` then DALI TensorFlow plugin package is built with prebuild plugin binaries inside. If PREBUILD_TF_PLUGINS is set to ``NO`` then the wheel is still built but not prebuild binaries are placed inside and the user needs to make sure that he has proper compiler version present (aligned with the one used to build present TensorFlow) so the plugin can be built during the installation of DALI TensorFlow plugin package. If is BUILD_TF_PLUGIN is set to ``NO`` PREBUILD_TF_PLUGINS value is disregarded. The default is ``YES``.
```
Comment (klecki) [inline docs/compilation.rst]:
```suggestion
* PREBUILD_TF_PLUGINS - Whether to prebuild DALI TensorFlow plugin. It should be used together with BUILD_TF_PLUGIN option. If both options are set to ``YES`` then DALI TensorFlow plugin package is built with prebuilt plugin binaries inside. If PREBUILD_TF_PLUGINS is set to ``NO`` then the wheel is still built but without prebuilding binaries - no prebuilt binaries are placed inside and the user needs to make sure that he has proper compiler version present (aligned with the one used to build present TensorFlow) so the plugin can be built during the installation of DALI TensorFlow plugin package. If is BUILD_TF_PLUGIN is set to ``NO`` PREBUILD_TF_PLUGINS value is disregarded. The default is ``YES``.
```

Comment (klecki) [reply docs/compilation.rst]: The diff doesn't color this well, but I would change some [pre]build into past participle form [pre]built.

File: docker/build.sh [hunk 4]
```diff
@@ -122,15 +131,18 @@ if [[ "$(docker images -q ${BUILDER} 2> /dev/null)" == "" || "$(docker images -q
...
-if [ "$BUILD_INHOST" = "YES" ]; then
+if [ "$BUILD_INHOST" == "YES" ]; then
     # build inside the source tree
-    docker run --rm -u 1000:1000 -v $(pwd):/opt/dali ${BUILDER} /bin/bash -c "mkdir -p /opt/dali/${DALI_BUILD_DIR} && \
+    docker run --rm -u $(id -u ${USER}):$(id -g ${USER}) -v $(pwd):/opt/dali ${BUILDER} /bin/bash -c "mkdir -p /opt/dali/${DALI_BUILD_DIR} && \
                                         cd /opt/dali/${DALI_BUILD_DIR} &&         \
                                         rm -rf /opt/dali/${DALI_BUILD_DIR}/nvidia* && \
+                                        ARCH=${ARCH}                              \
+                                        WHL_PLATFORM_NAME=${WHL_PLATFORM_NAME}    \
+                                        CUDA_TARGET_ARCHS=\"${CUDA_TARGET_ARCHS}\"\
```
Comment (klecki) [inline docker/build.sh]: No default for this? Is this intentional? Do we want to document it?

Comment (klecki) [reply docker/build.sh]: I know it is overwritten by CMake when it's empty.

Comment (JanuszL) [reply docker/build.sh]: It is intentional. If empty the defaults are used in scripts inside. The same as for most of the variables around (build_helper keeps these defaults in one place).

File: docs/compilation.rst (hunk 3 above)
Comment (JanuszL) [reply docs/compilation.rst]: Done in https://github.com/NVIDIA/DALI/pull/1407.

**Pull request merged**